File: apps/web/src/api/feedback.ts
import { AnalyticsApiError, formatAnalyticsApiError } from './analytics';

export interface FeedbackReference {
  code: string | null;
  nameVi: string | null;
}

export interface FeedbackWorkspaceItem {
  feedbackItemId: string;
  feedbackId: string;
  reportedAt: string;
  sourceSystem: string;
  contentMasked: string;
  location: { id: string | null; code: string | null; name: string | null };
  affectedChannelCodes: string[];
  currentClassification: {
    service: FeedbackReference | null;
    issue: FeedbackReference | null;
    sentiment: string | null;
    operationalSeverity: string | null;
    classificationState: string | null;
    projectionVersion: number | null;
  };
  status: string;
  analyticEligibility: string;
  parentItemId: string | null;
}

export interface FeedbackListFilters {
  projectId: string;
  dateFrom?: string;
  dateTo?: string;
  sourceSystem?: string;
  intakeChannelCode?: string;
  affectedChannelCode?: string;
  locationId?: string;
  serviceCode?: string;
  issueCode?: string;
  sentiment?: string;
  operationalSeverity?: string;
  customerLifecycleStageCode?: string;
  customerLifecycleStepCode?: string;
  touchpointCode?: string;
  hotspotId?: string;
  query?: string;
  limit?: number;
  offset?: number;
}

export interface FeedbackListResult {
  items: FeedbackWorkspaceItem[];
  total: number;
  limit: number;
  offset: number;
}

interface ApiReference { code: string | null; name_vi: string | null }
interface ApiFeedbackItem {
  feedback_item_id: string;
  feedback_id: string;
  reported_at: string;
  source_system: string;
  content_masked: string;
  location: { id: string | null; code: string | null; name: string | null };
  affected_channel_codes: string[];
  current_classification: {
    service: ApiReference | null;
    issue: ApiReference | null;
    sentiment: string | null;
    operational_severity: string | null;
    classification_state: string | null;
    projection_version: number | null;
  };
  status: string;
  analytic_eligibility: string;
  parent_item_id: string | null;
}

interface ApiListResponse {
  data: ApiFeedbackItem[];
  meta: { total: number; limit: number; offset: number };
}

import { getApiBaseUrl } from './config';

export const feedbackProjectId = import.meta.env.VITE_ANALYTICS_PROJECT_ID?.trim();

const toItem = (item: ApiFeedbackItem): FeedbackWorkspaceItem => ({
  feedbackItemId: item.feedback_item_id,
  feedbackId: item.feedback_id,
  reportedAt: item.reported_at,
  sourceSystem: item.source_system,
  contentMasked: item.content_masked,
  location: item.location,
  affectedChannelCodes: item.affected_channel_codes,
  currentClassification: {
    service: item.current_classification.service && { code: item.current_classification.service.code, nameVi: item.current_classification.service.name_vi },
    issue: item.current_classification.issue && { code: item.current_classification.issue.code, nameVi: item.current_classification.issue.name_vi },
    sentiment: item.current_classification.sentiment,
    operationalSeverity: item.current_classification.operational_severity,
    classificationState: item.current_classification.classification_state,
    projectionVersion: item.current_classification.projection_version,
  },
  status: item.status,
  analyticEligibility: item.analytic_eligibility,
  parentItemId: item.parent_item_id,
});

export const listFeedbackItems = async (filters: FeedbackListFilters): Promise<FeedbackListResult> => {
  const params = new URLSearchParams({ project_id: filters.projectId, limit: String(filters.limit ?? 100), offset: String(filters.offset ?? 0) });
  const fields: Record<Exclude<keyof FeedbackListFilters, 'projectId' | 'limit' | 'offset' | 'query'>, string> = {
    dateFrom: 'date_from',
    dateTo: 'date_to',
    sourceSystem: 'source_system',
    intakeChannelCode: 'intake_channel_code',
    affectedChannelCode: 'affected_channel_code',
    locationId: 'location_id',
    serviceCode: 'service_code',
    issueCode: 'issue_code',
    sentiment: 'sentiment',
    operationalSeverity: 'operational_severity',
    customerLifecycleStageCode: 'customer_lifecycle_stage_code',
    customerLifecycleStepCode: 'customer_lifecycle_step_code',
    touchpointCode: 'touchpoint_code',
    hotspotId: 'hotspot_id',
  };
  for (const [field, apiName] of Object.entries(fields)) {
    const value = filters[field as keyof typeof fields];
    if (value) params.set(apiName, value);
  }
  if (filters.query?.trim()) params.set('q', filters.query.trim());

  const response = await fetch(`${getApiBaseUrl()}/api/v1/feedback-items?${params.toString()}`);
  if (!response.ok) {
    let message = `Feedback API trả về lỗi ${response.status}.`;
    try {
      message = formatAnalyticsApiError((await response.json() as { detail?: unknown }).detail) ?? message;
    } catch {
      // Use the status message when the server did not return JSON.
    }
    throw new AnalyticsApiError(message, response.status);
  }
  const data = await response.json() as ApiListResponse;
  return { items: data.data.map(toItem), total: data.meta.total, limit: data.meta.limit, offset: data.meta.offset };
};

export const getFeedbackItem = async (feedbackItemId: string): Promise<FeedbackWorkspaceItem> => {
  const response = await fetch(`${getApiBaseUrl()}/api/v1/feedback-items/${feedbackItemId}`);
  if (!response.ok) throw new AnalyticsApiError(`Không tải được feedback item (${response.status}).`, response.status);
  return toItem((await response.json() as { data: ApiFeedbackItem }).data);
};

File: apps/api/schemas/feedback.py
"""Public, masked-only HTTP contracts for Feedback Workspace APIs."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class FeedbackAPIModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Reference(FeedbackAPIModel):
    code: str | None = None
    name_vi: str | None = None


class Location(FeedbackAPIModel):
    id: UUID | None = None
    code: str | None = None
    name: str | None = None


class CurrentClassification(FeedbackAPIModel):
    service: Reference | None = None
    issue: Reference | None = None
    journey_stage: Reference | None = None
    journey_step: Reference | None = None
    touchpoint: Reference | None = None
    sentiment: str | None = None
    operational_severity: str | None = None
    classification_state: str | None = None
    projection_version: int | None = None


class FeedbackItemData(FeedbackAPIModel):
    feedback_item_id: UUID
    feedback_id: UUID
    reported_at: datetime
    source_system: str
    content_masked: str
    location: Location
    affected_channel_codes: list[str]
    current_classification: CurrentClassification
    status: str
    analytic_eligibility: str
    parent_item_id: UUID | None = None


class FeedbackItemListMeta(FeedbackAPIModel):
    total: int = Field(ge=0)
    limit: int = Field(ge=1)
    offset: int = Field(ge=0)


class FeedbackItemListResponse(FeedbackAPIModel):
    data: list[FeedbackItemData]
    meta: FeedbackItemListMeta


class FeedbackItemDetailResponse(FeedbackAPIModel):
    data: FeedbackItemData


class SplitItemRequest(FeedbackAPIModel):
    item_text_masked: str = Field(min_length=1)
    symptom_detail: str | None = None
    location_id: UUID | None = None
    affected_channel_ids: list[UUID] = []


class SplitFeedbackItemRequest(FeedbackAPIModel):
    reason: str = Field(min_length=1)
    items: list[SplitItemRequest] = Field(min_length=2)


class SplitFeedbackItemResponse(FeedbackAPIModel):
    data: dict[str, object]


class UpdateFeedbackItemRequest(FeedbackAPIModel):
    service_code: str | None = None
    issue_code: str | None = None
    sentiment: str | None = None
    operational_severity: str | None = None
    analytic_eligibility: str | None = None
    location_id: UUID | None = None
    symptom_detail: str | None = None
    correction_reason: str | None = None
